rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java

```diff
@@ -312,6 +345,148 @@ public CompletableFuture<List<? extends CodeLens>> codeLens(CodeLensParams param
+            }
+        }
+
+        for (var msg : messages) {
+            client.showMessage(setMessageParams((IConstructor) msg));
+        }
+    }
+
+    private MessageParams setMessageParams(IConstructor message) {
+        var params = new MessageParams();
+        switch (message.getName()) {
+            case "warning": {
+                params.setType(MessageType.Warning);
+                break;
+            }
+            case "info": {
+                params.setType(MessageType.Info);
+                break;
+            }
+            default: params.setType(MessageType.Log);
+        }
+
+        var msgText = ((IString) message.get("msg")).getValue();
+        if (message.has("at")) {
+            var at = ((ISourceLocation) message.get("at")).getURI();
+            params.setMessage(String.format("%s (at %s)", msgText, at));
+        } else {
+            params.setMessage(msgText);
+        }
+
+        return params;
+    }
+
+    @Override
+    public void didRenameFiles(RenameFilesParams params, Set<ISourceLocation> workspaceFolders) {
+        Map<ILanguageContributions, List<FileRename>> byContrib =  bundleRenamesByContribution(params.getFiles());
+        for (var entry : byContrib.entrySet()) {
+            ILanguageContributions contrib = entry.getKey();
+            List<FileRename> renames = entry.getValue();
+
+            IList renameDocumentEdits = renames.stream().map(rename -> fileRenameToDocumentEdit(rename)).collect(VF.listWriter());
+
+            contrib.didRenameFiles(renameDocumentEdits)
+                .thenAccept(res -> {
+                    var edits = (IList) res.get(0);
+                    var messages = (ISet) res.get(1);
+                    showMessages(messages);
+
+                    if (edits.size() == 0) {
+                        return;
+                    }
+
+                    WorkspaceEdit changes = DocumentChanges.translateDocumentChanges(this, edits);
+                    client.applyEdit(new ApplyWorkspaceEditParams(changes)).thenAccept(editResponse -> {
+                    if (!editResponse.isApplied()) {
+                        throw new RuntimeException("didRenameFiles resulted in a list of edits but applying them failed"
+                            + (editResponse.getFailureReason() != null ? (": " + editResponse.getFailureReason()) : ""));
+                    }
+                });
+                })
+                .get()
+                .exceptionally(e -> {
+                    logger.catching(Level.ERROR, e.getCause());
+                    client.showMessage(new MessageParams(MessageType.Error, e.getCause().getMessage()));
+                    return null; // Return of type `Void` is unused, but required
+                });
+        }
+    }
+
+    private Map<ILanguageContributions, List<FileRename>> bundleRenamesByContribution(List<FileRename> allRenames) {
```

Comment:
since `prepareRename` is called by the workspace service for all file changes, we should first make sure, the current incoming call is actually for a file extension for which we have any handlers registered.

I think this function should filter out files we do not even care about. 

And this is only since we cannot predict, for the multiplexer, which file patterns would be more precise than all of the files.